### ISSUE TITLE
indentation messed up when using 'use' on top of PHP code

### ISSUE THREAD

**Issue by yaoweizhen** [emacs-php/php-mode#283]
It happened when adding "use" for namespace, but save the file and open again, the indentation works well.

**Comment by ejmr** [comment]
Something like this?

```php
use Foo\Bar;

class Baz
{
    private $blah;
}
```

**Comment by jwdunne** [comment]
I've noticed this too but I cannot remember how to reproduce, seems to happen at random - running M-x php-mode command fixes indentation, this is easier/faster than saving/reopening the file.

If I remember, I may have started to write the class, remembered the namespaces, added those in and the issue presented. I'll experiment and update.

**Comment by trashofmasters** [comment]
I also have been experiencing this problem for the past 6 months. I feel that sometimes triggering #281 makes this indentation problem more likely to occur. Normally it's fixed `M-x php-mode` or with `M-x revert-buffer`.

@ejmr Yes, as far as I can tell it happens with many files that declare one, or more, classes or functions. Although I'm not 100% positive it's related to the presence of `use` statements, as IIRC I've had a file declaring only a class affected by faulty indentation.

Is there anything I can do to collect data next time this errors occurs again?

**Comment by ejmr** [comment]
> Is there anything I can do to collect data next time this errors occurs again?

Please post as much of the code as you can so that we can build a unit test from it.  I've recently started work on a project using a lot of PHP and I have now run into this issue once myself, so I'm also trying to put together a test case we can use to replicate the problem.  Like you and others have said, `M-x revert-buffer` fixed the problem and I didn't run into it again when any other time I opened the file in which the problem occurred.  Which makes this a curious bug...

**Comment by trashofmasters** [comment]
I haven't been able to find a precise reproduction procedure, however, this morning I had an affected buffer with a class 100 lines long, here's a [gist with the content](https://gist.github.com/trashofmasters/00e3728d12fae1d647c3).

I found that using `M-: (setq debug-on-signal t)` and attempting to indent any part of the buffer reports unbalanced parenthesis and the following stack trace. 

I hope this helps

```
;; GNU Emacs 25.0.50.1
Debugger entered--Lisp error: (scan-error "Unbalanced parentheses" 112 113)
  scan-lists(7 1 -1)
  c-append-to-state-cache(7 113)
  c-parse-state-1()
  c-parse-state()
  c-guess-basic-syntax()
  c-indent-line()
  php-cautious-indent-line()
  indent-for-tab-command(nil)
  funcall-interactively(indent-for-tab-command nil)
  call-interactively(indent-for-tab-command)
  (progn (call-interactively beyond-yasnippet))
  (if beyond-yasnippet (progn (call-interactively beyond-yasnippet)))
  (let* ((yas-fallback-behavior (quote yas--fallback)) (yas-minor-mode nil) (beyond-yasnippet (yas--keybinding-beyond-yasnippet))) (yas--message 4 "Falling back to %s" beyond-yasnippet) (progn (or (or (null beyond-yasnippet) (commandp beyond-yasnippet)) (cl--assertion-failed (quote (or (null beyond-yasnippet) (commandp beyond-yasnippet))))) nil) (setq this-original-command beyond-yasnippet) (if beyond-yasnippet (progn (call-interactively beyond-yasnippet))))
  (cond ((eq yas-fallback-behavior (quote return-nil)) nil) ((eq yas-fallback-behavior (quote yas--fallback)) (error (concat "yasnippet fallback loop!\n" "This can happen when you bind `yas-expand' " "outside of the `yas-minor-mode-map'."))) ((eq yas-fallback-behavior (quote call-other-command)) (let* ((yas-fallback-behavior (quote yas--fallback)) (yas-minor-mode nil) (beyond-yasnippet (yas--keybinding-beyond-yasnippet))) (yas--message 4 "Falling back to %s" beyond-yasnippet) (progn (or (or (null beyond-yasnippet) (commandp beyond-yasnippet)) (cl--assertion-failed (quote (or ... ...)))) nil) (setq this-original-command beyond-yasnippet) (if beyond-yasnippet (progn (call-interactively beyond-yasnippet))))) ((and (listp yas-fallback-behavior) (cdr yas-fallback-behavior) (eq (quote apply) (car yas-fallback-behavior))) (let ((command-or-fn (car (cdr yas-fallback-behavior))) (args (cdr (cdr yas-fallback-behavior))) (yas-fallback-behavior (quote yas--fallback)) (yas-minor-mode nil)) (if args (apply command-or-fn args) (if (commandp command-or-fn) (progn (setq this-command command-or-fn) (call-interactively command-or-fn)))))) (t nil))
  yas--fallback()
  (if templates-and-pos (yas--expand-or-prompt-for-template (first templates-and-pos) (car (cdr templates-and-pos)) (car (cdr (cdr templates-and-pos)))) (yas--fallback))
  (let (templates-and-pos) (if (and yas-expand-only-for-last-commands (not (member last-command yas-expand-only-for-last-commands))) nil (setq templates-and-pos (if field (save-restriction (narrow-to-region (yas--field-start field) (yas--field-end field)) (yas--templates-for-key-at-point)) (yas--templates-for-key-at-point)))) (if templates-and-pos (yas--expand-or-prompt-for-template (first templates-and-pos) (car (cdr templates-and-pos)) (car (cdr (cdr templates-and-pos)))) (yas--fallback)))
  yas-expand()
  funcall-interactively(yas-expand)
  call-interactively(yas-expand nil nil)
  command-execute(yas-expand)
```

**Comment by trashofmasters** [comment]
Is anyone experiencing the issue using any form of semantic-php perhaps?

**Comment by ejmr** [comment]
@trashofmasters Thank you very much for the gist and trace output.  Unfortunately I do not get the same error---or any error---when I follow the same steps.  However, I am not using semantic-php.  So I will try that, because you may be correct and this problem could be the result of an error between how PHP Mode interacts together with semantic-php.

**Comment by trashofmasters** [comment]
I'll make an attempt to reproduce this without semantic-php. So far I known for sure this happens with either my own or Joris' implementation.

**Comment by redbeardcreator** [comment]
I don't have reproduce code but can say it happens to me without semantic-php. I am running flycheck with PHPCS and PHPMD, but I'm pretty sure I had this problem before flycheck. It is very sporadic and unpredictable.

**Comment by ejmr** [comment]
@redbeardcreator That has been my experience as well.  Thank you for helping with verifying that it can happen without semantic-php.

**Comment by trashofmasters** [comment]
@ejmr I think this issue could be linked to #281, in fact I could reproduce this issue by repeatedly causing the other. FWIW, the file only had one class and no use statements.

Here's the reproduction procedure using `emacs -Q` and PHP mode loaded with <kbd>M-x load-library</kbd>. 
1. Open a PHP buffer
2. After `<?php` press <kbd>M-j</kbd> one or multiple times
3. Delete all the trailing `\` characters
4. Select the buffer <kbd>C-x h</kbd> and indent it with <kbd>Tab</kbd>
5. Repeat until the indentation fails

With this procedure it took about 5 minutes before completely messing up the indentation of the buffer.

**Comment by trashofmasters** [comment]
Okay, my last procedure wasn't really accurate, but it happens that I found the trigger while working on something else.

Apparently even just `newline`, when used in the right place, can break indentation in the simplest of the php files:

```php
<?php echo "hello, world;"

```

<img src="https://cloud.githubusercontent.com/assets/30650/11918745/0c6b7dfa-a736-11e5-873e-8a475934d115.gif" width="468px"/>

As I understand it, if there's anything between the opening tag an the end of line, when the point is between the opening tag and some text pressing <kbd>Ret</kbd> or <kbd>M-j</kbd> will likely break the indentation. Although sometimes the first few newlines are added fine, repeating the sequence will increase the likelihood of completely messing up the buffer.

I discovered this whilst attempting to fix #281 by remapping `c-indent-comment-new-line` to another function, which prevents the unnecessary call to `c-backslash-region` by carrying out itself the indentation around php opening tags. The problem was fixed but the tests wouldn't pass because they triggered the indentation bug reported here.
